android/quest/src/main/java/org/smartregister/fhircore/quest/ui/report/measure/worker/MeasureReportMonthPeriodWorker.kt

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2021-2023 Ona Systems, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.smartregister.fhircore.quest.ui.report.measure.worker
+
+import android.content.Context
+import androidx.hilt.work.HiltWorker
+import androidx.work.CoroutineWorker
+import androidx.work.WorkerParameters
+import com.google.android.fhir.FhirEngine
+import com.google.android.fhir.workflow.FhirOperator
+import dagger.assisted.Assisted
+import dagger.assisted.AssistedInject
+import kotlinx.coroutines.withContext
+import org.smartregister.fhircore.engine.configuration.ConfigType
+import org.smartregister.fhircore.engine.configuration.ConfigurationRegistry
+import org.smartregister.fhircore.engine.configuration.report.measure.MeasureReportConfiguration
+import org.smartregister.fhircore.engine.data.local.DefaultRepository
+import org.smartregister.fhircore.engine.util.DispatcherProvider
+import org.smartregister.fhircore.engine.util.extension.SDF_YYYY_MM_DD
+import org.smartregister.fhircore.engine.util.extension.firstDayOfMonth
+import org.smartregister.fhircore.engine.util.extension.formatDate
+import org.smartregister.fhircore.engine.util.extension.lastDayOfMonth
+import org.smartregister.fhircore.engine.util.extension.loadCqlLibraryBundle
+import org.smartregister.fhircore.engine.util.extension.retrievePreviouslyGeneratedMeasureReports
+import org.smartregister.fhircore.engine.util.extension.today
+import org.smartregister.fhircore.quest.data.report.measure.MeasureReportRepository
+import timber.log.Timber
+
+@HiltWorker
+class MeasureReportMonthPeriodWorker
+@AssistedInject
+constructor(
+  @Assisted val context: Context,
+  @Assisted workerParams: WorkerParameters,
+  val fhirEngine: FhirEngine,
+  private val fhirOperator: FhirOperator,
+  val defaultRepository: DefaultRepository,
+  val dispatcherProvider: DispatcherProvider,
+  private val measureReportRepository: MeasureReportRepository,
+  val configurationRegistry: ConfigurationRegistry,
+) : CoroutineWorker(context, workerParams) {
+
+  /**
+   * This method always sets the start and end date of the month to be the first and last day of the
+   * current month. If you are using this worker to generate monthly reports it enqueue a new worker
+   * at least every 24 hours.
```

Comment:
I'm still not sure this is correct, need to think about this a bit more to see if there's an edge case that misses data from the previous month when the date changes.

Reply:
I think there is actually still a problem, since it’s using “today” to figure out the current month instead of the date it was enqueued. This is what I’m thinking, if it’s the last day of the month, you run the worker, and enqueue another worker for tomorrow. Then you collect more data. That new data you collected will never appear in the previous months measure report, because by the time the enqueued measure report runs it will tomorrow, in the next month, and that run will be for the next month.

Reply:
One (simple?) solution is to have `today` be set when the worker is enqueued, instead of when the worker is run. then when you enqueue a worker on the last day of the month, and proceed to collect some data, when that worker runs it will cover all the data for the previous month, during which it was enqueued.

This also seems to match expected behavior... better?

Reply:
created this issue to track this, https://github.com/opensrp/fhircore/issues/2567